### PR TITLE
use abstract superclass for param errors

### PR DESCRIPTION
**Pull request by repomaa**
this allows type safe rescuing

### CODE REVIEW

**Comment by vladfaust** [comment]
So, what's the main purpose on this PR? To introduce abstract `ParamError` class? I agree on that, however the code seems to become much more difficult.

**Comment by repomaa** [comment]
Hm yeah, it was an attempt to DRY things up a little. I can revert back to the way things were before if you like.

**Comment by vladfaust** [comment]
Yes, please keep the things simpler. Also I'd like you to conform with Angular Commit Styling (as seen at https://github.com/vladfaust/prism/commits/master). I suggest the commit message to be as following:

```
feat: abstract `ParamError` class

`InvalidParamTypeError`, `ParamNotFoundError`, `InvalidParamError` and `ProcError` all now extend `ParamError` class

**Comment by repomaa** [comment]
will do!

**Comment by repomaa** [comment]
Better?

**Comment by vladfaust** [comment]
Yes, that's better, but still not ideal. I also would like to change how `Action::Params` handle errors. Could you please give me an access to your PR? https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by repomaa** [comment]
allow edits from maintainers is already checked.

**Comment by repomaa** [comment]
I left the detail param to be `String?` because of the off chance that someone would call the initializer and pass `nil` to it so nothing would break. I don't like it though. The default could be passed to `@detail : String` directly in the params. `@message` was a bit of an ill choice for a instance var for a class that inherits from `Exception` because it gets overwritten in the `Exception` constructor ;)

File: src/prism/params/errors.cr [hunk 1]
```diff
@@ -1,68 +1,86 @@
 module Prism::Params
+  abstract class ParamError < Exception
+    getter param
+
+    def initialize(@param : AbstractParam)
+      super(build_message)
+    end
+
+    def path
+      param.path + [param.name]
```
Comment (vladfaust) [inline src/prism/params/errors.cr]: Change to `param.path.join(param.name)` to avoid array initialization

Comment (vladfaust) [inline src/prism/params/errors.cr:10]: Change to `param.path.push(param.name)` to avoid array initialization

**Review by vladfaust — CHANGES_REQUESTED** [review]

File: src/prism/params/errors.cr [hunk 2]
```diff
@@ -1,68 +1,86 @@
...
+        path:    path.join(" > "),
+        detail: detail,
+      }
     end
   end
 
-  class ProcError < Exception
-    getter param, message
+  class ProcError < ParamError
+    getter detail : String
```
Comment (vladfaust) [inline src/prism/params/errors.cr:70]: Change to `details`, it's more appropriate here

**Review by vladfaust — CHANGES_REQUESTED** [review]

**Review by vladfaust — CHANGES_REQUESTED** [review]
`Action::Params` and `Channel::Params` should be changed as well to rescue the abstract class

File: src/prism/params/errors.cr (hunk 1 above)
Comment (repomaa) [reply src/prism/params/errors.cr:10]: doesn't that mutate `param.path`?

**Comment by vladfaust** [comment]
You'd also want to close #19

**Comment by repomaa** [comment]
Close #19? How come? I don't deal with any error codes here.

**Comment by vladfaust** [comment]
If `Action::Params` is changed, #19 can be closed alongside

**Comment by repomaa** [comment]
I think that should be a separate PR since it's a completely separate issue.

File: src/prism/params/errors.cr (hunk 1 above)
Comment (vladfaust) [reply src/prism/params/errors.cr:10]: It definitely does, I'm an idiot.

**Pull request closed without merging**